doc: Zephyr out-of-tree module.yml examples

gregshue opened the pull request:
This commit adds module.yaml examples which support the
zephyr directory structure beneath <module-root>/zephyr,
and describes optional (out-of-tree) module scope.

Signed-off-by: Gregory Shue <gregory.shue@legrand.us>

Commit:
doc: Zephyr out-of-tree module.yml examples

This commit adds module.yml examples which support the
zephyr directory structure beneath <module-root>/zephyr,
and describes downstream (supplemental) module scope.

Signed-off-by: Gregory Shue <gregory.shue@legrand.us>

gregshue commented:
@carlescufi , please also review this (or invite a delegate).

carlescufi commented on doc/guides/modules.rst:
```diff
@@ -18,6 +18,19 @@ Zephyr depends on several categories of modules, including:
 - Inter-Process Communication (IPC) libraries
 
+Zephyr also supports extension of the functionaly with out-of-tree modules.
+Out-of-tree modules will likely be their own manifest repository during
```
This should link to the west documentation regarding topologies, so that users are informed of what "be their own manifest repository" means

carlescufi commented on doc/guides/modules.rst:
```diff
@@ -153,8 +175,32 @@ Build settings supported in the :file:`module.yml` file are:
+
+.. code-block:: yaml
+
+   build:
```
Perhaps we could use another folder instead of `zephyr`, to be more explicit?

mbolivar-nordic replied:
> Perhaps we could use another folder instead of `zephyr`, to be more explicit?

I want to bikeshed the chosen alternative 'zephyrous' so hard, but I won't :smile:

carlescufi commented on doc/guides/modules.rst:
```diff
@@ -153,8 +175,32 @@ Build settings supported in the :file:`module.yml` file are:
 
-Example of a :file:`module.yaml` file containing additional roots, and the
-corresponding file system layout.
+Example of a :file:`module.yaml` file containing additional roots
```
```suggestion
Example of a :file:`module.yml` file containing additional roots
```

carlescufi commented on doc/guides/modules.rst:
```diff
@@ -153,8 +175,32 @@ Build settings supported in the :file:`module.yml` file are:
+       ├── dts
+       └── soc
+
+Example of a :file:`module.yaml` file containing additional roots located
```
```suggestion
Example of a :file:`module.yml` file containing additional roots located
```

carlescufi commented on doc/guides/modules.rst:
```diff
@@ -186,8 +232,26 @@ To execute both tests and samples available in modules, the Zephyr test runner
+
+.. code-block:: yaml
 
+   build:
```
Same comment here, should we use a different folder instead of `zephyr`?

tejlmand commented on doc/guides/modules.rst:
```diff
@@ -32,6 +32,21 @@ Zephyr depends on several categories of modules, including but not limited to:
 - File Systems
 - Inter-Process Communication (IPC) libraries
 
+Zephyr also supports extension of functionality with out-of-tree modules.
```
This out-of-tree is no more, no less out of tree of the [zephyr workspace](https://docs.zephyrproject.org/latest/guides/zephyr_cmake_package.html#zephyr-application-structure) when comparing with Zephyr manifest described modules.

Why not use the term: `Zephyr extra modules` as we already describe how those can be specified:
https://docs.zephyrproject.org/latest/guides/modules.html#integrate-modules-in-zephyr-build-system
and then describe that those extra modules can be provided by a custom / downstream manifest file ?

gregshue replied:
@tejlmand As I recall, the Zephyr documentation consistently(?):

- Refers to the modules that the `zephyr` repository depends upon as `external`, and these modules have a `modules/Kconfig.*` file within the `zephyr` repository (due to limits of Kconfig)

- Refers to the modules that the `zephyr` repository does **not** depend upon as `out of tree`, and all configuration, source, and build instructions are in the external module itself. 

I have been keeping the following questions in mind as I have wrestled with the concept and implementation of modules:

- The definition of 'module' is currently independent of 'manifest'. Does that matter? (N)

- IF the Zephyr Build System team decided it would be much easier if the build system discovered the `zephyr` repository in the same way it discovers modules, would the definition of 'module' still fit? (Y)

- IF the `zephyr` repository had an alternate manifest file which also included all downstream modules hosted at `zephyrproject-rtos`, do all the terms and definitions still fit? (?)

I really want the term 'module' == `zephyr/module.yml`.  The mechanism is already independent of dependence from `zephyr` repository, location of config/glue code, hosting service, open source vs. commercial vs. private. I want the definition of the term to reflect that scope. This will avoid confusion about scope and will track capabilities as the module.yml schema is extended. I think this will greatly simplify communication now and in the future. It does mean we need distinct terms for each of the above sub-cases.  

I have been trying to stick with the existing terms, but it may be time for them to be overhauled.  Unfortunately that means some files get renamed. I would prefer to not use the phrase 'zephyr extra modules' so we don't imply an artificial and unnecessary coupling between the CMake variables and the amount of dependence the `zephyr` repository has on the module.

Perhaps we should adopt the following?

-  'upstream module' for those modules the `zephyr` repository depends upon
-  'downstream module' for those modules the `zephyr` repository does **not** depend upon
-  'Zephyr Project module` for those modules managed/maintained at `zephyrproject-rtos`

jackrosenthal replied:
FWIW, our team uses zephyr modules without west entirely. So for us, module really does mean `zephyr/module.yml`.

tejlmand replied:
> @tejlmand As I recall, the Zephyr documentation consistently(?):

@gregshue Thanks for the answer, but i'm not addressing the `module` term.
What you describe as `module` is correct, what I'm addressing, is the term `out-of-tree`.

> Refers to the modules that the zephyr repository depends upon as external, and these modules have a modules/Kconfig.* file within the zephyr repository (due to limits of Kconfig)

The reason for moving the Kconfig files back to Zephyr, was to avoid warnings: https://github.com/zephyrproject-rtos/zephyr/pull/16512.

> Refers to the modules that the zephyr repository does not depend upon as out of tree

If you notice #16512, then those modules are optional, so you cannot say Zephyr depends upon the modules referenced in `module/Kocnfig.*`. Hence, saying that an `out-of-tree` module is a module Zephyr does not depend upon may lead to the false impression that Zephyr depends on all the other modules (which Zephyr doesn't, as some of them are optional).

Certain features of Zephyr or hw might depend on them, but if not using those features or hw, you can easily leave them out.
Example, if I don't need stm, then I can leave out https://github.com/zephyrproject-rtos/hal_stm32, or if not using Nordic, I can leave out: https://github.com/zephyrproject-rtos/hal_nordic.

Also, the build code of the external module are for now, still located in the external module.

> I really want the term 'module' == zephyr/module.yml. 

which is correct. You seem to have missed the intention of my comment.

> I would prefer to not use the phrase 'zephyr extra modules' so we don't imply an artificial and unnecessary coupling between the CMake variables and the amount of dependence the zephyr repository has on the module.

So the fact that the guide describes:
> The build system variable ZEPHYR_MODULES is a CMake list of absolute paths to the directories containing Zephyr modules.

means we cannot use the term `Zephyr modules` because it is also used by CMake ?

You seem to be under the impression that CMake uses the argument `ZEPHYR_EXTRA_MODULES`.
It doesn't, it simply passes on that piece of information to: https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/zephyr_module.py
here:
https://github.com/zephyrproject-rtos/zephyr/blob/e7df25c848c6ab54bc81430605454f91ca204623/cmake/zephyr_module.cmake#L33-L36

and then `zephyr_module.py` processes the information in `module.yml` and determine if that module.yml contains additional Kconfig, CMake, SoC / Board / DTS roots to include in the build system.
Just like it uses `west` to fetch manifest modules.
I therefore think Zephyr extra modules is the closest term that actually describes your intention.

Maybe if we cleaned up the module description (as you are already in process of doing :+1: ), then we could describe:
- Zephyr Modules 
- Zephyr Extra Modules

and then describe how to specify `Zephyr Extra Modules` in a custom / downstream manifest, or providing those with CMake.

Zephyr Module is not exclusive for CMake, so why should the term Zephyr Extra Module be that ?

> - 'upstream module' for those modules the zephyr repository depends upon
> - 'downstream module' for those modules the zephyr repository does not depend upon

I would refrain from using `upstream` / `downstream` as some users might be experiencing Zephyr for the first time when using a downstream Zephyr based SDK, like nRF Connect SDK.
In such cases it can become less clear what upstream / downstream refers to.

Finally, `upstream module` can still be optional, so I would not say that Zephyr depends on them.

gregshue replied:
@tejlmand Thank you for your detailed feedback. I'm glad to read that my description of `module` is correct.

Please note that I didn't originate the term out-of-tree.  It seems from your proposal that cleaning up usages and moving/renaming files/samples/etc is an acceptable result.

Actually, I was looking at the interface to the build system independent of the current implementation. Whether the build system **currently** implements logic around ZEPHYR_EXTRA_MODULES in CMake statements, python, ruby, or ksh does not matter to the users of the build system. Similarly, since the build system does not require west to be installed, the manifest concept (however integrators choose to manage it) and locations for optional west manifest files are critical.

Reading the current modules.rst documentation and looking into `cmake/zephyr_module.cmake` and `scripts/zephyr_module.py`, it seems ZEPHYR_EXTRA_MODULES is reserved for a specific use case described in modules.rst, while ZEPHYR_MODULES is the standard way of passing any/all module paths into the Zephyr build system. I think we need to not conflict with the ZEPHYR_EXTRA_MODULES use case either directly or through inference and avoid using 'extra' adjective elsewhere.

Here is a list of the various conditions for which we need (hopefully unambiguous) phrases:

1. **'Zephyr module'**
    Any repository having `zephyr/module.yml`  (agreeable)
2. **'Zephyr manifest module'**
    Any Zephyr module holding one or more west manifest files.
3. **'Zephyr Project module'** ?
    Any Zephyr module hosted at `zephyrproject-rtos`
4. **'Zephyr Project manifest module'** ?
    Any Zephyr manifest module hosted at `zephyrproject-rtos`
5. ?
    Any Zephyr non-manifest module hosted at `zephyrproject-rtos`
6. **'Zephyr default module'** ?
    Any Zephyr module in `zephyr` repository default manifest
7. ?
    Any Zephyr module with `zephyr/manifest/Kconfig.<module>`

I like "optional" for any module that doesn't have to be cloned to build the minimal Zephyr system (e.g. samples/hello_world for native_posix). Given that criteria, are there any modules which are NOT optional?

tejlmand replied:
> It seems from your proposal that cleaning up usages and moving/renaming files/samples/etc is an acceptable result.

> Reading the current modules.rst documentation and looking into cmake/zephyr_module.cmake and scripts/zephyr_module.py, it seems ZEPHYR_EXTRA_MODULES is reserved for a specific use case described in modules.rst, while ZEPHYR_MODULES is the standard way of passing any/all module paths into the Zephyr build system. I think we need to not conflict with the ZEPHYR_EXTRA_MODULES use case either directly or through inference and avoid using 'extra' adjective elsewhere.

Yes, and I see the term `ZEPHYR_EXTRA_MODULES` has not been precise enough described from the start.
https://github.com/zephyrproject-rtos/zephyr/pull/13067



It was original intended as a way to use vanilla Zephyr, and then specify additional repos / folders that should also be considered as Zephyr modules.
Up until now, there has not been a need for distinguish between Zephyr modules, and then (Extra/downstream/custom) Zephyr modules, and when using `west` all projects are in the manifest are treated as a potential Zephyr module.

With the future desire to move Zephyr module glue code (CMake code) into the Zephyr repository (ref: Process WG), then it might be that it becomes important to be able to distinguish between Zephyr module and then (Extra/downstream/custom) Zephyr modules.
In such case, this distinction could be important, and then terminology becomes important.
CMake arguments should reflect the terminology we decide upon.

> Whether the build system currently implements logic around ZEPHYR_EXTRA_MODULES in CMake statements, python, ruby, or ksh does not matter to the users of the build system. 


Except for the fact, that if we describe something as `Zephyr Downstream Module`, then the build argument should reflect that as `-DZEPHYR_DOWNSTREAM_MODULES`.
That way it is much easier for a user to relate terminology and build arguments.

> Similarly, since the build system does not require west to be installed, the manifest concept (however integrators choose to manage it) and locations for optional west manifest files are critical.

Not sure I understand what you mean. `west` is optional, so Zephyr modules should not depend on `west`.
As of now, it appears to me that the user guide part of Zephyr Modules are describing this nicely.

---------
And to the interesting part:

> 1.  'Zephyr module'
>      Any repository having zephyr/module.yml (agreeable)

Yes, unless we already now should consider the fact that in close future we might actually need to distinguish between a Zephyr Module, and then a user defined module.

> 2.   'Zephyr manifest module'
>       Any Zephyr module holding one or more west manifest files.
 
Not sure I understand why this is important.
Are there any situations where it is important to be aware if a module also holds a manifest file ?
and why is that important ?

> 3.   'Zephyr Project module' ?
>       Any Zephyr module hosted at zephyrproject-rtos

That could actually be a good term for the Zephyr Project Module, then there is no need for specifying Zephyr Extra Modules.
You can then specify either `ZEPHYR_MODULES` or `ZEPHYR_PROJECT_MODULES` (the latter will have glue code in Zephyr repo).
Initial feeling on this is good.

>   4.   'Zephyr Project manifest module' ?
>    Any Zephyr manifest module hosted at zephyrproject-rtos

Same as 2.

>  5.  ?
>     Any Zephyr non-manifest module hosted at zephyrproject-rtos

Partly same as 2. But how is this module different from 3. ?

>  6.   'Zephyr default module' ?
>        Any Zephyr module in zephyr repository default manifest

How is this module different from 3. ?
If a Zephyr module is hosted at zephyrproject-rtos, then it should be in the manifest, otherwise there is no reason to host it.
Am I missing a use-case ?

>   7.   ?
>      Any Zephyr module with zephyr/manifest/Kconfig.<module>

? Not sure I understand.

> I like "optional" for any module that doesn't have to be cloned to build the minimal Zephyr system (e.g. samples/hello_world for native_posix). Given that criteria, are there any modules which are NOT optional?

All modules today are optional:
```
$ cmake -DBOARD=native_posix -DZEPHYR_MODULES="dummy" -GNinja  -Bbuild samples/hello_world/
Including boilerplate (Zephyr repository): /projects/github/zephyr/zephyr/cmake/app/boilerplate.cmake
-- Application: /projects/github/zephyr/zephyr/samples/hello_world
-- Zephyr version: 2.4.99 (/projects/github/zephyr/zephyr)
...
-- Generating done
-- Build files have been written to: /projects/github/zephyr/zephyr/build
$ cat build/zephyr_modules.txt  # <--- This files is empty
$ ninja -Cbuild 
ninja: Entering directory `build'
[1/81] Preparing syscall dependency handling

[81/81] Linking C executable zephyr/zephyr.elf
```
Note: Just discovered that using `-DZEPHYR_MODULES=""` actually results in default modules being loaded :disappointed: . Therefore using `dummy` to disable all modules.

---------

While reading / re-reading this PRs in context of whole documentation and replying, then I started to notice that part of this may be due to my old comment here:
https://github.com/zephyrproject-rtos/zephyr/pull/27185#pullrequestreview-464415109
> Would like to see a split of the docs into the existing Modules (External Projects) user guide, and then a dedicated Modules (External Projects) contribution guide.

that some of your intentions might be to make it clearer for contributors on how to contribute a new Zephyr module, and/or maintain existing ones.

And my focus originates more on the usages / integration part.
Of course those two things should be aligned, especially when it comes to terminology, but it also shows why this document should probably be divided into two different parts, `Usage` and `Contribution`.

gregshue replied:
> With the future desire to move Zephyr module glue code (CMake code) into the Zephyr repository (ref: Process WG), then it might be that it becomes important to be able to distinguish between Zephyr module and then (Extra/downstream/custom) Zephyr modules.

The future arrived a couple months ago. :-)



> >    Whether the build system currently implements logic around ZEPHYR_EXTRA_MODULES in CMake statements, python, ruby, or ksh does not matter to the users of the build system.
> 
> Except for the fact, that if we describe something as Zephyr Downstream Module, then the build argument should reflect that as -DZEPHYR_DOWNSTREAM_MODULES.
  That way it is much easier for a user to relate terminology and build arguments.

We have the same goal for the user experience, which led me to be comfortable with -DZEPHYR_MODULES as it already carries the paths to **all** modules (and parallels the definition of 'Zephyr module'). So far I have found no need to pass the different types of modules to the build system as separate lists. Have you? 

> Yes, unless we already now should consider the fact that in close future we might actually need to distinguish between a Zephyr Module, and then a user defined module.

We are there. See https://github.com/bacnet-stack/bacnet-stack and https://github.com/gregshue/oot_module_example. What term would you prefer to describe the publicly available modules provided outside of `zephyrproject-rtos`?

> Are there any situations where it is important to be aware if a module also holds a manifest file ?
and why is that important ?

It is important to the integrators, module developers, and those writing documentation, but not to the build system. The presence of a manifest file indicates that the `zephyr` repository is independent of this module, but this module is dependent on the `zephyr` repository and possibly other non-Zephyr Project modules. This affects how and where integrators triage and report issues, and how developers work. It also matters if the non-Zephyr Project modules provide multiple manifest files. This cascades over into Zephyr Project repository naming patterns.

> If a Zephyr module is hosted at zephyrproject-rtos, then it should be in the manifest, otherwise there is no reason to host it.
Am I missing a use-case ?

Yes, you are missing the use case in issue #28438.

> >        Any Zephyr module with zephyr/manifest/Kconfig.
> 
> ? Not sure I understand.

There are some modules that have no manifest/Kconfig.<module> file. (e.g. fatfs). There are others which do (e.g. nordic), and there are some files in modules/Kconfig.* which do not have a corresponding module.yml file (e.g. mcux, imx, vega). Unless we can migrate these Kconfig.* files out to the modules, we will need some phrase to describe them in discussion and documentation.

tejlmand replied:
> > With the future desire to move Zephyr module glue code (CMake code) into the Zephyr repository (ref: Process WG), then it might be that it becomes important to be able to distinguish between Zephyr module and then (Extra/downstream/custom) Zephyr modules.
>
> The future, err .. time when it became important, arrived a couple months ago. :-)

Yes, but the work itself has not been done yet (hence: the future).
Meaning, we still don't know if this work will need the ability to distinguish between the type of module, in order to load the files correctly.
Or how to handle the use case where a downstream project forks a Zephor hosted module, and then patch it.
Would users have to patch BOTH zephyr (with gluecode) and the module, or will it be enough to patch just the module, examples of such modules:
https://github.com/nrfconnect/sdk-mcuboot (making it a Zephyr module)
https://github.com/nrfconnect/sdk-mcumgr (just patches)

> So far I have found no need to pass the different types of modules to the build system as separate lists. Have you?

Not sure. when moving build logic into Zephyr repository, there could be a need for distinguish between the type of modules, so that downstream users may still use their own fork of a Zephyr hosted module.
Either by patching the module, or by having code in their manifest repository having higher precedence than the Zephyr repository code.
Exactly what is needed / how to handle it / implement, is yet to be decided.

> > Are there any situations where it is important to be aware if a module also holds a manifest file ?
> >    and why is that important ?
>
> It is important to the integrators, module developers, and those writing documentation, but not to the build system. The presence of a manifest file indicates that the zephyr repository is independent of this module, but this module is dependent on the zephyr repository and possibly other non-Zephyr Project modules. This affects how and where integrators triage and report issues, and how developers work. It also matters if the non-Zephyr Project modules provide multiple manifest files. This cascades over into Zephyr Project repository naming patterns.

Still not sure I see the issue.
A repository can be:
- A Zephyr module
- A Manifest repository
- Both

But manifest repository and Zephyr module are orthogonal to each other, so they can be described and handled independently.

If looking at a given repository as a Zephyr module, then why does it matter whether this module also host a manifest file ?
The module may describe arch/soc/board/dts roots, contain Kconfig files, or CMake build files.
It has no knowledge of manifest files, and don't care for those.
So the fact that this repository may also function as a manifest repository seems irrelevant when describing the module functionality.
The build system use ZEPHYR_MODULES or `west list` as a list of candidates to parse as Zephyr modules, besides that, there is no link to manifest or `west`.

If looking at a given repository as a manifest repository, then why does it matter if this repository is also a Zephyr module ?
The manifest file describe projects to clone/update with `west update`.
`west` and the manifest file has no relation to Zephyr modules.
Just like they have no relationship to what language is used in source code of that module, C, C++, Python, Java ?
So the fact that this repository may also function as a Zephyr module seems irrelevant when describing the manifest functionality.

So why describe manifest and Zephyr module together ?
Keep them independent, cause they are independent.

> > If a Zephyr module is hosted at zephyrproject-rtos, then it should be in the manifest, otherwise there is no reason to host it.
> >  Am I missing a use-case ?
>
> Yes, you are missing the use case in issue #28438.

but those Zephyr modules will still be in a Zephyr hosted manifest file.
I don't think we should have different naming conventions depending on whether a Zephyr module is part of manifest repo A, B, or C.
Today we only have `zephyrproject-rtos/zephyr` as manifest project.
With #28438 there will be one more, and in future, there could even be more, as example, an incubated manifest repository, for projects not fully allowed into Zephyr mainline.
We may refer to those manifest repositories differently, but seen from Zephyr module perspective, they should all follow the same naming conventions and they may move from one manifest file to another, or be present in multiple.

Also I don't see a need for mixing Zephyr modules with manifest repo description, because of #28438.

> >    ? Not sure I understand.
> 
> There are some modules that have no manifest/Kconfig. file. (e.g. fatfs). There are others which do (e.g. nordic), and there are some files in modules/Kconfig.* which do not have a corresponding module.yml file (e.g. mcux, imx, vega). Unless we can migrate these Kconfig.* files out to the modules, we will need some phrase to describe them in discussion and documentation.

A Zephyr module is not required to have a Kconfig or CMakeLists.txt file, that is optional.
You can create a Zephyr module purely for tests, or a Zephyr module containing additional board roots.

In the case you describe with `Kconfig.imx/mcux/vega`, then it seems to be simply bad naming.
Technically those files are NOT part of the Zephyr module, but was placed here, instead of in the module, for the reason I linked earlier: #16512.
So in fact they are just ordinary Kconfig files in the Zephyr repo. (and I understand the confusion this can lead to).

When enhancing the Zephyr module handling, so that module glue code can be located outside the repo containing the module itself, then this should be cleaned up, in order to be consistent.

So rather than having a term for this, it should be cleaned up.

This seems to turn into a bigger discussion, than what I initially anticipated when addressing the `out-of-tree` term :laughing:

gregshue replied:
> Meaning, we still don't know if this work will need the ability to distinguish between the type of module, in order to load the files correctly.

By extension of the existing pattern, any metadata needed by the build system about a module is to exist in `zephyr/module.yml`. The schema for that will grow over time as we identify needed metadata, but the scope of the content is already defined.

> Or how to handle the use case where a downstream project forks a Zephor hosted module, and then patch it.
  Would users have to patch BOTH zephyr (with gluecode) and the module, or will it be enough to patch just the module,   examples of such modules:
  https://github.com/nrfconnect/sdk-mcuboot (making it a Zephyr module)
  https://github.com/nrfconnect/sdk-mcumgr (just patches)

I have already experienced this when Nordic maintained a public fork of Zephyr (driving the need for my application repository to have alternate manifest files. The interfaces between the repositories doesn't move. If the issue is in the interface, then all repositories using that interface get patched. If the issue is not in the interface, then only the repositories with the issues get patched. This holds true no matter where the glue code is. In both cases the user will need want to create a clone of the repository and apply the patch to their own long-term branch rather than the upstream main development branch. When the upstream repository has a fix for the issue then the user chooses which implementation to use. (This is just standard-operating-procedure. :-))

> Not sure. when moving build logic into Zephyr repository, there could be a need for distinguish between the type of modules, so that downstream users may still use their own fork of a Zephyr hosted module.
  Either by patching the module, or by having code in their manifest repository having higher precedence than the Zephyr repository code.
  Exactly what is needed / how to handle it / implement, is yet to be decided.

Once a user forks a `zephyrproject-rtos` repository they can do whatever they want. All we can do is provide recommendations. I have already been working against forks of all the Zephyr Project hosted repositories on other Git servers. I have already updated and privately patched the forks to work around issues. This helped me recognize the need for policies/guidance includes prefixes on branch names and tags (to avoid collisions).

> So the fact that this repository may also function as a manifest repository seems irrelevant when describing the module functionality.

The difference is relevant to:
- Module release processes (which is now independent of `zephyr`)
- Zephyr Project documentation
- Zephyr Project CI pipeline (since the module build needs to use a different manifest file)
- Where issues are tracked(?)

> The build system use ZEPHYR_MODULES or west list as a list of candidates to parse as Zephyr modules, besides that, there is no link to manifest or west.

There is no link to the manifest **file** or west, but there is still an upstream/downstream relationship between the repositories that affects how the downstream module is managed.

> ...
   So the fact that this repository may also function as a Zephyr module seems irrelevant when describing the manifest functionality.

That is true with respect to the source build system. That is no longer true for document generation along with the other points mentioned above.

> So why describe manifest and Zephyr module together ?
  Keep them independent, cause they are independent.

When I created a downstream module I needed to develop it in isolation against a known, tracked version of Zephyr. This meant the downstream module development environment had one or more distinct manifest that needed to be tracked and maintained. (I actually simultaneously had three in the same branch: current Zephyr master, latest Zephyr release, prior Zephyr release.)

When I integrated this downstream module into an internal device, the prj.conf for the device was in another repository with its own superset manifest. This one is independent of the module.

> ...
> but those Zephyr modules will still be in a Zephyr hosted manifest file.

You just changed from "the manifest" to "a manifest".  Take it the next step and identify *which* manifest file in *which* repository.  :-)

> Today we only have zephyrproject-rtos/zephyr as manifest project.
  With #28438 there will be one more, and in future, there could even be more, as example, an incubated manifest repository, for projects not fully allowed into Zephyr mainline.

Exactly. We will likely end up with a manifest-only repository to verify that the #28438 repository content works as a module with an external manifest. In addition, we may choose to move one or more of our existing upstream modules to be a downstream module (e.g. Civetweb).

BUT the phrase "not allowed into Zephyr mainline" is misleading. These will be repositories that are part of the Zephyr Project mainline functionality. They are purposefully kept separated from the `zephyr (RTOS)` repository to help verify the essential functionality of extensibility. 

> We may refer to those manifest repositories differently, but seen from Zephyr module perspective, they should all follow the same naming conventions and they may move from one manifest file to another, or be present in multiple.

**Absolutely!** The question here is how they are named. We have already evolved to have HAL modules begin with the hal_ prefix. Remember, the purpose of the naming convention is to aid automation and convey meaning to people. It may be useful for Zephyr Project CI automation to have a rule that says "All zpmm_* repositories are to be run separately with each manifest matching zephyr/manifests/zp_ci/*/west.yml"

**Do we similarly want a distinct prefix for downstream modules? For example modules?**

> A Zephyr module is not required to have a Kconfig or CMakeLists.txt file, that is optional.

Oh! I missed that and assumed we needed to use empty files instead. 😆

> When enhancing the Zephyr module handling, so that module glue code can be located outside the repo containing the module itself, then this should be cleaned up, in order to be consistent.

I have already demonstrated that and am actively depending on the Zephyr module handling **currently** supports glue code outside the `zephyr` repository. No enhancement of the build system is needed. (I did need west 0.8.01a though.) Ticket #28438 is primarily about enhancing the Zephyr Project CI coverage to verify the functionality is maintained for each of the extension hooks.

> This seems to turn into a bigger discussion, than what I initially anticipated when addressing the out-of-tree term 

The good news is the vast majority of the things we are discussing have a natural solution that follows the existing patterns. (Sometimes the existing patterns take some work to recognize though e.g. ZEPHYR_MODULES carries **all types** of modules, not just the upstream ones.)

Once we settle on module terms then the name of the repository for #28438 is determined. The rest follows the pattern of existing contribution processes and should flow smoothly and quickly.

tejlmand replied:
> > Meaning, we still don't know if this work will need the ability to distinguish between the type of module, in order to load the files correctly.
>                                                                                                   
> By extension of the existing pattern, any metadata needed by the build system about a module is to exist in zephyr/module.yml. The schema for that will grow over time as we identify needed metadata, but the scope of the content is already defined.
                                                                                                    
I'm not talking about the zephyr/module.yml.                                                        
But even today, with #16512, the design of Zephyr modules actually broke partly.                    
Cause all the code in `modules/Kconfig.*` is not really Zephyr modukle glue code, it's just regular Kconfig code.
What still lacks is correct integration into Zephyr module handling.                                
                                                                                                    
There is no proper handling / safe guarding of whether a module is actually available.              
                                                                                                    
> >    Or how to handle the use case where a downstream project forks a Zephor hosted module, and then patch it.
> >    Would users have to patch BOTH zephyr (with gluecode) and the module, or will it be enough to patch just the module, examples of such modules:
> >    https://github.com/nrfconnect/sdk-mcuboot (making it a Zephyr module)                        
> >    https://github.com/nrfconnect/sdk-mcumgr (just patches)                                      
>                                                                                        
> I have already experienced this when Nordic maintained a public fork of Zephyr (driving the need for my application repository to have alternate manifest files. The interfaces between the repositories doesn't move. If the issue is in the interface, then all repositories using that interface get patched. If the issue is not in the interface, then only the repositories with the issues get patched. This holds true no matter where the glue code is. In both cases the user will need want to create a clone of the repository and apply the patch to their own long-term branch rather than the upstream main development branch. When the upstream repository has a fix for the issue then the user chooses which implementation to use. (
>                                                                                                   
> This is just standard-operating-procedure. :-)                                                    
                                                                                                    
For the simple case, yes.                                                                           
But let's say I have an older Zephyr, then I want to update a module to Module-latest, because it fixes some critical bug.
I cannot simply update the module, cause the glue code has changed, so now I need to decide if I want to also start patching Zephyr, because glue code changes are needed.
In this situation, it might be beneficial to be able to simply take the latest glue code and place it directly in the module, and even if outdated glue code is also available in Zephyr-old release, then the code in the module takes precedence.
Or maybe even be able to place glue code in alternative repo, to avoid patching zephyr project repo completely.
                                                                                                    
> >  Not sure. when moving build logic into Zephyr repository, there could be a need for distinguish between the type of modules, so that downstream users may still use their own fork of a Zephyr hosted module.
    Either by patching the module, or by having code in their manifest repository having higher precedence than the Zephyr repository code.
    Exactly what is needed / how to handle it / implement, is yet to be decided.                    
>                                                                                            
> Once a user forks a zephyrproject-rtos repository they can do whatever they want.                 
                                                                                                    
Yes, but we should try to create the infrastructure in such a way so user can hook into the system, without having to fork and patch.
If a module is identified as downstream hosted, then the build infrastructure could allow for looking up glue code in another repo.
That would be an example of where distiguishing between Zephyr hosted / non-hosted module could be valuable.
                                                                                                    
                                                                                                    
> > So the fact that this repository may also function as a manifest repository seems irrelevant when describing the module functionality.
>                                                                                                
> The difference is relevant to:                                                                    
>                                                                                                   
> -  Module release processes (which is now independent of zephyr)                                  
                                                                                                    
Not sure what you mean by Module release processes ?                                                
Zephyr is not releasing modules, so are you refering to upmerging of Externally handled modules, such as mbedtls, openamp, etc. ?
If we talk about upmerges and hence PRs to a Zephyr hosted module, then it should be stated:        
- All PRs to a Zephyr module must be accompanied by corresponding PRs to all Zephyr hosted manifest repositories that includes the Zephyr module, where revision is updated tp the form `pull/<no>head`.
                                                                                                    
That way it doesn't matter if the module is also a manifest repository.                             
If the module is used by another manifest repo, then it must have a corresponding PR, regardless whether it is also a manifest repo itself.
                                                                                                    
>    Zephyr Project documentation                                                                   
                                                                                                    
Project documentation is generated from the manifest repository.                                    
And if that manifest repository include docs from Zephyr module(s), then I see no problem in doing so. 
I still fail to see why the fact that a module repo can also be a manifest repo is special.         
                                                                                                    
                                                                                                    
>    Zephyr Project CI pipeline (since the module build needs to use a different manifest file)        
                                                                                                    
That would be covered by the requirement to have a manifest update.                                 
                                                                                                    
>    Where issues are tracked(?)                                                                    
                                                                                                    
The manifest project where the issue was identified.                                                
So if some finds an issue when using `zephyr`, the report the issue towards Zephyr, which then might spawn additional issues elsewhere.
Like this chain: #26114 --> https://github.com/OpenAMP/open-amp/pull/213 and https://github.com/OpenAMP/libmetal/pull/119
and then waiting for later upmerge.                                                                 
                                                                                                    
With this approach, then a module doesn't need to care.                                             
The manifest project where the issue was identified, is where the issue should be reported.         
If multiple projects are affected, the issue can be cloned.                                         
When a fix has been merged to the module, and manifest files in various projects has been updated to include latest version of the module, then the issue can be closed.
Just like today. I see no reason to change that, even if Zephyr starts to have multiple manifest repositories.
                                                                                                       
Similar to what is done at Nordic, where might be reported against our local Zephyr fork. If it is identified to also exist in Zephyr, then an issue is also created in Zephyr, and fixed there.
                                                                                                       
> When I created a downstream module I needed to develop it in isolation against a known, tracked version of Zephyr. This meant the downstream module development environment had one or more distinct manifest that needed to be tracked and maintained. (I actually simultaneously had three in the same branch: current Zephyr master, latest Zephyr release, prior Zephyr release.)
>                                                                                                      
> When I integrated this downstream module into an internal device, the prj.conf for the device was in another repository with its own superset manifest. This one is independent of the module.

                                                                                                    
To me, this seems to be overcomplicating the initial work of a new module.                          
You could have used `-DZEPHYR_EXTRA_MODULE=/path/to/my/new/module` together with CMake or west.     
https://docs.zephyrproject.org/latest/guides/modules.html#integrate-modules-in-zephyr-build-system  
                                                                                                    
and then just switched between Zephyr vX.Y.Z / vO.P.Q/ master and tested the module.                
I think that if the intention is to just test a given module at revision A with n-versions of Zephyr, then the manifest approach you just described is not the best way.
It could be you have found new use case that should be considered as improvement, but I still don't see why Zephyr module should be described together with manifest.
                                                                                                    
> We have already evolved to have HAL modules begin with the hal_ prefix.                           
                                                                                                    
And such a naming pattern is good, cause it describe the purpose of the repo, that it contains hal code.
But prefixing the repo because it contains a manifest file is not good, cause that would mean repos would change name if at later point it should not hold a manifest file (or the other way around).
                                                                                                    
> It may be useful for Zephyr Project CI automation to have a rule that says "All zpmm_* repositories are to be run separately with each manifest matching zephyr/manifests/zp_ci/*/west.yml"
                                                                                                    
I think that if we start to have so many repositories that we would benefit from automation here, then I'm concerned people will loose track of where to update manifest files.
                                                                                                    
> > When enhancing the Zephyr module handling, so that module glue code can be located outside the repo containing the module itself, then this should be cleaned up, in order to be consistent.
>                                                                                                   
> I have already demonstrated that and am actively depending on the Zephyr module handling currently supports glue code outside the zephyr repository. No enhancement of the build system is needed.
                                                                                                    
I assume you mean: `outside the module repository` ?                                                
That code won't work as expected. Of course you can place CMake code anywhere in Zephyr you like, and have it to refer to the module repo, but that won't integrate the module correctly into Zephyr.
It will just be regular CMake code, so can we agree to reserve Zephyr module glue code term to what is actually part of the Zephyr module infrastructure ?
A couple of caveats:                                                                                
- `ZEPHYR_CURRENT_MODULE_DIR` won't be set when processing the CMake code, so code inside a module depending on this variable will not work.
  Also the Zephyr extension command: `zephyr_library_amend` will not work.                          
- Same for `ZEPHYR_CURRENT_CMAKE_DIR`                                                               
- `ZEPHYR_<module_name>_CMAKE_DIR` will not be set. (This could be the reason you needed an empty CMakeLists.txt file, if other parts of code depended on this variable)
- The depends keyword will not work properly.                                                       


But the original comment I made regarding `out-of-tree` term has still not been addressed.

gregshue replied:
@tejlmand Thanks for the insights. :-)

> There is no proper handling / safe guarding of whether a module is actually available.

True, for modules that are upstream of `zephyr`. We already do have one for those that are downstream of `zephyr` (which is the main focus of this issue).

> But let's say I have an older Zephyr, then I want to update a module to Module-latest, because it fixes some critical bug.
I cannot simply update the module, cause the glue code has changed, so now I need to decide if I want to also start patching Zephyr, because glue code changes are needed.

In the case above I'll assume the interface between the two repositories changed. Otherwise the repository(s) can be changed independently to address the issue(s). Addressing processes for managing changes in upstream modules is out-of-scope for this issue.

> In this situation, it might be beneficial to be able to simply take the latest glue code and place it directly in the module, and even if outdated glue code is also available in Zephyr-old release, then the code in the module takes precedence.

How do we control the library precedence used by the Zephyr build system?


> Yes, but we should try to create the infrastructure in such a way so user can hook into the system, without having to fork and patch.

Absolutely, which is why I am focusing on downstream modules.

> If a module is identified as downstream hosted, then the build infrastructure could allow for looking up glue code in another repo.

There is no need to change the current build infrastructure. It is already finding the glue code in the downstream module.

> Not sure what you mean by Module release processes ?
Zephyr is not releasing modules, so are you refering to upmerging of Externally handled modules, such as mbedtls, openamp, etc. ?

Tying the release of Zephyr to the release of Zephyr Project-managed **downstream** modules seems unnecessary, so I **assumed** Zephyr Project-managed **downstream** modules would eventually have their own release numbers and release processes. 

> If we talk about upmerges and hence PRs to a Zephyr hosted module, then it should be stated:

We are primarily talking about changes to Zephyr hosted downstream modules. In either case, your statements about needing to update mainfests is true, and is already addressed in many cases through the 'import' keyword in west manifest files.

> To me, this seems to be overcomplicating the initial work of a new module.

When I created a module out of an existing open source project I found there were three things I cared about. 

- Could I use it in another manifest?
- How do I develop this with the minimal dependencies?
- What is the easiest way to tie into the existing Zephyr build system that is consistent with the design?

The last question led me to **avoid** having other contributors to my module require special configurations with ZEPHYR_EXTRA_MODULES.

> >    I have already demonstrated that and am actively depending on the Zephyr module handling currently supports glue code outside the zephyr repository. No enhancement of the build system is needed.
> 
> I assume you mean: outside the module repository ?

No. I have not tried to address upstream modules with glue code in the `zephyr` repository. That is out-of-scope for this issue.

Pulling this back to the scope of this issue/ticket, it seems that you prefer:
- using a term for downstream modules other than the already existing term "out-of-tree".
- not having a naming pattern for Zephyr Project repositories that indicates which are also manifest repositories

(I ignored the ZEPHYR_MODULES vs ZEPHYR_EXTRA_MODULES because all we can provide is a recommendation on usage.)

Have I missed anything else related to this ticket?  What do you see as the pending questions?

mbolivar-nordic replied:
This discussion is too long to read, but here are my two cents:

"Out of tree" has always, without exception, meant "outside the zephyr repository" to me.

It has never had anything to do with whether or not there is an official blessing from the Zephyr Project itself for any functionality in the module.

So "out-of-tree modules" is simply redundant, and "in-tree module" would be an oxymoron.

gregshue replied:
@mbolivar-nordic Thank you for your comments. You will want to read the discussion in order to make them relevant.

BTW, the `zephyr` repository subdirectory `samples/application_development/out_of_tree_driver/hello_world_module` is technically an in-tree module as you define it.

tejlmand reviewed (changes requested):
Thanks for this. Good to get this described.

I think the out-of-tree term is a bit confusing, as the Zephyr manifest described modules are also out-of-tree, when comparing to the Zephyr repository, but all are still inside the Zephyr workspace.

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -32,6 +32,21 @@ Zephyr depends on several categories of modules, including but not limited to:
 
+Zephyr also supports extension of functionality with out-of-tree modules.
+These are modules downstream of the Zephyr repository and may be managed
+outside of the Zephyr Project.  Out-of-tree modules will likely be their
```
This "will likely be" is your use case, I know, but I don't think it's appropriate for this documentation, unless you have some hard data showing that it is most likely.

mbolivar-nordic commented on doc/guides/modules.rst:
```diff
@@ -669,9 +788,9 @@ created by the project team and initialized with basic information that would
 allow submitting code to the module project following the project contribution
 guidelines.
 
-If a module is maintained as a fork of another project on Github, the Zephyr module
-related files and changes in relation to upstream need to be maintained in a
-special branch named ``zephyr``.
+If a module is maintained as a fork of another project on Github, the Zephyr
+module related files and changes in relation to upstream need to be maintained
+in a special branch named ``zephyr``.
```
This looks like clutter in the diff which doesn't change any text.

mbolivar-nordic reviewed (commented):
Nice examples!

I do have a few whinges about what I perceive as improper usage and editorializing, but I'm not putting a -1 on this to avoid blocking.

tejlmand commented:
> This discussion is too long to read, but here are my two cents:

@mbolivar-nordic No worries, I commented simply on the out-of-tree usage myself, and that for reasons that are little unclear, even to me, turned into a longer discussion that had nothing to do with the out-of-tree term.
So regarding the out-of-tree term, you didn't miss anything.

gregshue commented:
@tejlmand @mbolivar-nordic Agreed, No worries.

> ... for reasons that are little unclear, even to me, turned into a longer discussion that had nothing to do with the out-of-tree term.

Perhaps I can clear that up:

    This thread is about choosing terms and definitions that clarify the breadth of extensibility already enabled by `zephyr/module.yml` and supported by the Zephyr build system. 

This may require a paradigm shift, as evidenced by the "But what about ...?" and "But that means ..." type of Q&A above. :-)

gregshue commented:
I have extended the glossary with Zephyr module related terms and definitions to work against. Once we settle on the replacement phrase for 'out-of-tree' then I will update modules.rst.

dleach02 commented:
This PR has gone quiet but seems like a useful PR to complete. I've removed the stale tag to prevent it from being closed but the author should consider resolving the conflicts and change requests to see if this can get merged... or close it if they don't want to pursue it

gregshue commented:
@dleach02, thank you for noticing this. #28438 is working through alignment on the scope of a module, whether up-stream of Zephyr or down-stream of it. I expect to get back to this once a repository is created for that issue.  (Any documentation edits before that are likely to change and lead to confusion.)

Do you think it is beneficial to make these documentations before then?

tejlmand commented on doc/glossary.rst:
```diff
@@ -96,4 +96,35 @@ Glossary of Terms
+   Zephyr manifest module
+      A Zephyr module which also contains one or more west manifest files.
+
+   Zephyr module
+      A YAML file named :file:`zephyr/module.yml` and all the content it
+      references, which define the extensions to be hooked into the Zephyr
+      infrastructure.
```
imo we consider the whole project a Zephyr module when it contains a `zephyr/module.yml` file, and not just the file itself.

The file is just used to identify that this project is a Zephyr module, and the content of the file defines how it should be integrated, as example, is it a DTS, board, soc root ? should it be included in the build flow (CMakeLists/Kconfig) ? should it integrate into twister.

gregshue replied:
@tejlmand I started with your impression, then asked if it still made sense for repositories like project_chip and bacnet-stack, which have substantial content for non-Zephyr platforms and environments. It seems the content not reachable by the Zephyr glue code is beyond any Zephyr-module policies. I just didn't want developers in other Open Source projects to think all of their repository needed to comply with Zephyr policies.

Would you still like this changed?

tejlmand replied:
I think we need something in between. Cause I find the current description too limiting.

Personally, I would still refer to the project as a Zephyr module, although not all of the code in there may be used by Zephyr.
Just like https://github.com/zephyrproject-rtos/mbedtls is a Zephyr module, but as the code is developed by ARM, then the code itself will not comply with Zephyr policies.

gregshue replied:
@tejlmand Suggestions welcome, and I can align with a more generic definition. Since we can imply a valid `zephyr/module.yml` from repos with both `zephyr/CMakeLists.txt` and `zephyr/Kconfig` (and/or deprecate it), how about:

"A repository with a valid `zephyr/module.yml`" ?

tejlmand commented on doc/glossary.rst:
```diff
@@ -96,4 +96,35 @@ Glossary of Terms
+      and default manifest file are aware of.
+
+   :makevar:`ZEPHYR_MODULES`
+      The Zephyr build system CMake variable through which a list paths to
+      all Zephyr modules may be provided.
```
what about:
```suggestion
     The Zephyr build system CMake variable through which a list of paths to
     Zephyr modules that should be included in the current build.
```

You could have `n`-Zephyr modules available, but decide to only include `m` of those in a particular build of projectA, while including `l` of them when building projectB.

In this case, the user probably will use a variant of 2. in this description: https://docs.zephyrproject.org/latest/guides/modules.html#without-west

tejlmand commented on doc/glossary.rst:
```diff
@@ -96,4 +96,35 @@ Glossary of Terms
+
+   :makevar:`ZEPHYR_EXTRA_MODULES`
+      The Zephyr build system CMake variable through which paths to
+      Zephyr modules may be provided.  This is provided to support module
+      development tangential to how :makevar:`ZEPHYR_MODULES` is being used.
```
or it can be used in combination to `west`, where `west` knows about the manifest defined modules, but you have additional local modules or private repos (as example for testing or confidential code) that you want to include in a build.

Could be, as example, when working with the source for a closed source stack.

gregshue replied:
Good to know. Please suggest replacement text if you feel that needs to be captured.

tejlmand replied:
will do some thinking.

tejlmand commented on doc/guides/modules.rst:
```diff
@@ -529,10 +587,14 @@ requires the following folder structure:
 
    <module-root>
+   ├── zephyr/module.yml
    ├── arch
    ├── boards
    ├── dts
-   └── soc
+   ├── soc
+   ├── ...
+   ├── CMakeLists.txt
+   └── Kconfig
```
As this code block now contains both files and folders, should we add `/` to the end of folders to indicate what is what ?
```suggestion
   ├── zephyr/module.yml
   ├── arch/
   ├── boards/
   ├── dts/
   ├── soc/
   ├── ...
   ├── CMakeLists.txt
   └── Kconfig
```

tejlmand commented on doc/guides/modules.rst:
```diff
@@ -542,23 +604,82 @@ Twister (Test Runner)
+       ├── module.yml
+       ├── CMakeLists.txt
+       ├── Kconfig
+       ├── arch
+       ├── boards
+       ├── dts
+       ├── samples
+       ├── soc
+       └── tests
```
Same comment regarding highlighting of folders.
```suggestion
       ├── arch/
       ├── boards/
       ├── dts/
       ├── samples/
       ├── soc/
       └── tests/
```

tejlmand reviewed (commented):
@gregshue thanks for the update, looks good.

A couple of minor suggestions when re-reading the updated docs.

gregshue committed:
Update doc/glossary.rst

Co-authored-by: Torsten Tejlmand Rasmussen <tejlmand@users.noreply.github.com>

gregshue committed:
Update doc/guides/modules.rst

Co-authored-by: Torsten Tejlmand Rasmussen <tejlmand@users.noreply.github.com>

Commit:
Update doc/guides/modules.rst

(Github web interface would not accept change submitted on a deleted
line, so Torsten's changes were manually recreated and committed.)

Co-authored-by: Torsten Tejlmand Rasmussen <tejlmand@users.noreply.github.com>
Signed-off-by: Gregory Shue <gregory.shue@legrand.us>
